### DISCUSSION TITLE
Shelly1 Rules Switch Timing Problem

### DISCUSSION THREAD

**Discussion by dieckfr** [arendst/Tasmota#18915] [Q&A]
I want to use a flashed shelly 1 for a Cow Brusch... There is a Reed Relais for Start and Stop after a Time...  I want the Shelly to work as an TimeRelais with delayed Power off. Here my Rule:

Rule1 on switch1#state=1 do backlog  power1 on; RuleTimer1 0   endon on switch1#state=0 do RuleTimer1 30 endon on Rules#Timer=1 do Power1 off endon

This works well when there is a slow change by Switch1 hardware inputt, but several times thiss occures:

0:14:02.352 RUL: SWITCH1#STATE=0 performs "RuleTimer1 30"
20:14:02.358 RSL: RESULT = {"T1":30,"T2":270,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
20:14:02.595 RSL: RESULT = {"T1":0,"T2":270,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}

This force the Drive to run endless... The Switch is off for a long time, but the shelly ignores the "off" state... and i dont understand whats happend there in the log whith my used Rules, i think there is no way for this...

Maybe is my fault, please let me know #10190 

My Workaround at the Moment is Rule2... but i want to know what there is happend...

Rule2 on Power1#State=1  do RuleTimer2 600 endon on Power1#State=0 do  RuleTimer2 0 endon on Rules#Timer=2 do Power1 0 endon

Thank a lot froom a Cow Friend :)

**Comment by sfromis** [comment]
It is not at all clear what happens, as what you post as "this occurs" includes RuleTimer2 which you only mentioned as an additional workaround, not included in your "situation". To get better debugging, do not have more rules than what you hoped to be workable, activate `Weblog 3` for details, and test. Then the log should tell a story about what happens. Also you are not clear about what is "slow change" or what you mean when it is no longer "slow change".

Your setting of SwitchMode1 and 2 makes a major difference to how the switch inputs work, and you mentioned nothing about that.

**Reply by dieckfr** [reply]
Dear sfromis,
thank you for your reply. The Switchmode makes no difference, i testet positiv and negative logic. Always the same. Slow means, that there is maybe a second between changes off the Switch hardware input. Rule2 as workaround:

#Rule2 on Power1#State=1  do RuleTimer2 300 endon on Power1#State=0 do  RuleTimer2 0 endon on Rules#Timer=2 do Power1 0 endon


I post the Wrong debug, sorry about that... And i have to say that i do my best in wording in english... Iam not a natural speaker :)

Here is the log with Weblog 3 whitout Rule2:

15:11:44.637 RUL: SWITCH1#STATE=0 performs "RuleTimer1 60"
15:11:44.639 SRC: Rule
15:11:44.642 CMD: Grp 0, Cmd 'RULETIMER', Idx 1, Len 2, Pld 60, Data '60'
15:11:44.647 RSL: RESULT = {"T1":60,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
15:11:46.093 RUL: SWITCH1#STATE=1 performs "backlog  power1 on; RuleTimer1 0"
15:11:46.095 SRC: Rule
15:11:46.098 CMD: Grp 0, Cmd 'BACKLOG', Idx 1, Len 23, Pld -99, Data 'power1 on; RuleTimer1 0'
15:11:46.107 SRC: Backlog
15:11:46.110 CMD: Grp 0, Cmd 'POWER', Idx 1, Len 2, Pld 1, Data 'on'
15:11:46.114 RSL: RESULT = {"POWER1":"ON"}
15:11:46.116 RSL: POWER1 = ON
15:11:46.330 SRC: Backlog
15:11:46.333 CMD: Grp 0, Cmd 'RULETIMER', Idx 1, Len 1, Pld 0, Data '0'
15:11:46.337 RSL: RESULT = {"T1":0,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
15:11:46.395 RUL: SWITCH1#STATE=0 performs "RuleTimer1 60"
15:11:46.397 SRC: Rule
15:11:46.400 CMD: Grp 0, Cmd 'RULETIMER', Idx 1, Len 2, Pld 60, Data '60'
15:11:46.405 RSL: RESULT = {"T1":60,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
15:11:47.898 RUL: SWITCH1#STATE=1 performs "backlog  power1 on; RuleTimer1 0"
15:11:47.900 SRC: Rule
15:11:47.903 CMD: Grp 0, Cmd 'BACKLOG', Idx 1, Len 23, Pld -99, Data 'power1 on; RuleTimer1 0'
15:11:47.912 SRC: Backlog
15:11:47.915 CMD: Grp 0, Cmd 'POWER', Idx 1, Len 2, Pld 1, Data 'on'
15:11:47.919 RSL: RESULT = {"POWER1":"ON"}
15:11:47.921 RSL: POWER1 = ON
15:11:48.128 RUL: SWITCH1#STATE=0 performs "RuleTimer1 60"
15:11:48.130 SRC: Rule
15:11:48.132 CMD: Grp 0, Cmd 'RULETIMER', Idx 1, Len 2, Pld 60, Data '60'
15:11:48.137 RSL: RESULT = {"T1":60,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
15:11:48.347 SRC: Backlog
15:11:48.351 CMD: Grp 0, Cmd 'RULETIMER', Idx 1, Len 1, Pld 0, Data '0'
15:11:48.355 RSL: RESULT = {"T1":0,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}


The drive runs for ever... Hope this debug can help you understanding?

**Reply by sfromis** [reply]
You were supposed to *read* the log, to understand what happens, instead of just copy/pasting to here.

One thing to take note of in the log is that when using `Backlog`, is that you're seeing the default inter-command delay of 200 ms, meaning that if something else happens in that interval, you may be shooting yourself in the foot. In general, I'd recommend having "next RuleTimer" command as the first in a backlog sequence, and maybe use `Backlog0` instead of `Backlog` to avoid that delay.

**Reply by dieckfr** [reply]
Dear sfromis,
backlog0 was the answer to the Problem! The 200ms delay was the Problem whith the Switch! Thanks you a lot!

**Comment by dieckfr** [comment]
Dear Stromis...
i give the Projekt to my buddy, and that is not the Solution we need... Maybe you have an produktive Idea for me... At Home every thing seems to work... but a litte bad feeling was there because it works 95% not every Time... I tested also the tasmota flashed shelly whitout Network, and everything seems fine... (I know that Tasmota driven things can make trouble in Network, and then they are mostly hard to get on IP, everything ist very slow... But so what... the thing i want to know is... is it possible to set a tasmota driven HArdware to a complete stupid thing that only give an Impuls a definied longer Time in on Mode... A simple signal thats comming without an rule pulses a lot of time whitout sense, should give an Output whith an definied Time... Thats all... My Favorit bonus, thats why i want to use Tasmota, was an AP or when needet a choise to read out hours of use, an how often a signal came... The Relay is used whit an Cow brush, there is an Signal over a sensor and and 2 Magnets when a Cow give the brusch a move, but whe the cow push it to high there is no signal, but retriggering is need when the brusch cames a little bit down, an there is a signal again... Have you any Idea? An is it possible to set Rules to ROM whats powersave whitout hardcoding? 
I like Tasmota and want to learn more what is possible, but my head jungelling whith the possibility to Code 5 lines, and push it to the hardware hardwired... but a shelly is an hardware that should work, and i want to solve the problem...

**Reply by sfromis** [reply]
Not sure how much I understand of that, but one Tasmota feature is PulseTime, meaning that Tasmota will keep an output on for the specified period after a input signal to turn on. As the interval will be restarted/extended for each new input signal, the output will be kept on until no input signal appeared for the selected interval. This works well with inputs like a PIR, but also with other types of inputs indicating activity.
